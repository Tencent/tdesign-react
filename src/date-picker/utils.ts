import chunk from 'lodash/chunk';
import { DateValue } from './type';
import { Styles } from '../common';

/**
 * next Month
 * @since version V1.0.0
 * @param { Date } date
 * @returns {Date}
 */
export const nextMonth = (date: Date) => {
  const isPassYear = date.getMonth() === 11;
  const newMonth = isPassYear ? 0 : date.getMonth() + 1;
  const year = isPassYear ? date.getFullYear() + 1 : date.getFullYear();
  return new Date(year, newMonth, 1);
};

/**
 * prev Month
 * @since version V1.0.0
 * @param {Date} date
 * @returns {Date}
 */
export const prevMonth = (date: Date) => {
  const passYear = date.getMonth() === 0;
  const newMonth = passYear ? 11 : date.getMonth() - 1;
  const year = passYear ? date.getFullYear() - 1 : date.getFullYear();
  return new Date(year, newMonth, 1);
};

/**
 * dateIndexOf
 * @since version V1.0.0
 * @param {Date} date
 * @param {Date[]} arrDates
 * @returns {Number}
 */
export function dateIndexOf(date: Date, arrDates: DateValue[]): number {
  const curDate = new Date(date);
  const curTime = curDate.setHours(0, 0, 0, 0);
  const selectedDates = arrDates.map((d: Date) => {
    const dt = new Date(d);
    const dtime = dt.setHours(0, 0, 0, 0);
    return dtime;
  });

  // console.log('curDate', selectedDates, curDate, selectedDates.indexOf(curTime));

  return selectedDates.indexOf(curTime);
}

/**
 * Empty value
 * @since version V1.0.0
 * @param mixedVar
 * @returns {Boolean}
 * @example
 * // example 1: empty(null)
 * // returns 1: true
 * // example 2: empty(undefined)
 * // returns 2: true
 * // example 3: empty([])
 * // returns 3: true
 * // example 4: empty({})
 * // returns 4: true
 * // example 5: empty({'fn' : function () {} })
 * // returns 5: false
 */
export function empty(mixedVar: any): mixedVar is undefined | null | false | '' | {} {
  let undef: undefined;
  const emptyValues = [undef, null, false, ''];
  for (let i = 0, len = emptyValues.length; i < len; i++) {
    if (mixedVar === emptyValues[i]) {
      return true;
    }
  }

  if (typeof mixedVar === 'object') {
    return Object.keys(mixedVar).length < 1;
  }

  return false;
}

/**
 * To Dashed from Camel Case
 * @since version V1.0.0
 * @param {String} strCamelCase
 * @returns {string}
 * @example
 * toDash('strCamelCase') === 'str-camel-case'
 */
export function toDash(strCamelCase: string): string {
  return String(strCamelCase)
    .split(/(?=[A-Z])/)
    .join('-')
    .toLowerCase();
}

export function setStyles(el: HTMLElement | SVGElement, styles: Styles) {
  Object.entries(styles).forEach(([prop, val = null]) => {
    if (empty(val)) {
      el.style.removeProperty(prop);
      return;
    }
    const [value, pri = ''] = String(val).split('!');

    el.style.setProperty(toDash(prop), value, pri);
  });
}

/**
 * 首字母大写
 * @param {String} str 目标字符串
 * @returns {String}
 */
export function firstUpperCase(str: string): string {
  if (!str) return str;
  return str[0].toUpperCase().concat(str.substring(1, str.length));
}

interface DateObj {
  year: number;
  month: number;
}

/**
 * 返回指定年、月的第一天日期
 * @param {Object} { year, month }
 * @returns {Date}
 */
function getFirstDayOfMonth({ year, month }: DateObj): Date {
  return new Date(year, month, 1);
}

/**
 * 返回指定年、月的天数
 * @param {Object} { year, month }
 * @returns {Number}
 */
function getDaysInMonth({ year, month }: DateObj): number {
  return new Date(year, month + 1, 0).getDate();
}

/**
 * 返回指定年、月的最后一天日期
 * @param {Object} { year, month }
 * @returns {Date}
 */
function getLastDayOfMonth({ year, month }: DateObj): Date {
  return new Date(year, month, getDaysInMonth({ year, month }));
}

function isSameYear(date1: Date, date2: Date) {
  return date1.getFullYear() === date2.getFullYear();
}

function isSameMonth(date1: Date, date2: Date) {
  return isSameYear(date1, date2) && date1.getMonth() === date2.getMonth();
}

function isSameDate(date1: Date, date2: Date) {
  return isSameMonth(date1, date2) && date1.getDate() === date2.getDate();
}

/**
 * 是否是某法范围内的日期，精确到日
 * @param {Date} value 目标日期
 * @param {Object} { start, end } 范围
 * @returns {Boolean}
 */
function isBetween(
  value: { getFullYear: () => number; getMonth: () => number; getDate: () => number },
  { start, end }: { start: any; end: any },
): boolean {
  const date = new Date(value.getFullYear(), value.getMonth(), value.getDate());

  const startTime = new Date(start.getFullYear(), start.getMonth(), start.getDate());
  const endTime = new Date(end.getFullYear(), end.getMonth(), end.getDate());

  return startTime <= date && endTime >= date;
}

/**
 * 比较两个日期对象的时间戳
 * @param {Date} date1 日期1
 * @param {Date} date2 日期2
 * @returns {Number} 返回 date1.getTime() - date2.getTime() 的差值
 */
function compareAsc(date1: { getTime: () => any }, date2: Date): number {
  const d1 = date1.getTime();
  const d2 = date2.getTime();

  if (d1 < d2) return -1;
  if (d1 > d2) return 1;
  return 0;
}

/**
 * 比较两个 Date 是否是同一天 或则 同一月 或则 同一年
 * @param {Date} date1 比较的日期
 * @param {Date} date2 比较的日期
 * @param {String} type 比较类型，默认比较到『日』 date|month|year
 * @returns {Boolean}
 */
export function isSame(date1: Date, date2: Date, type = 'date'): boolean {
  const func = {
    isSameYear,
    isSameMonth,
    isSameDate,
  };
  return func[`isSame${firstUpperCase(type)}`](date1, date2);
}

export function outOfRanges(d: Date, min: any, max: any) {
  return (min && compareAsc(d, min) === -1) || (max && compareAsc(d, max) === 1);
}

/**
 * @returns {Date} 当天零点的日期对象
 */
export function getToday(): Date {
  const now = new Date();
  return new Date(now.getFullYear(), now.getMonth(), now.getDate(), 0, 0, 0);
}

/**
 * 返回日期对象的年、月、日、小时、分钟、秒、12小时制标识
 * @param {Date} date
 * @returns {Object}
 */
export function getDateObj(date: Date) {
  if (!(date instanceof Date)) {
    throw new Error('date must be Date object');
  }
  return {
    year: date.getFullYear(),
    month: date.getMonth(),
    date: date.getDate(),
    hours: date.getHours(),
    minutes: date.getMinutes(),
    seconds: date.getSeconds(),
    meridiem: date.getHours() > 11 ? 'PM' : 'AM',
  };
}

/**
 * 设置日期对象的时间部分
 * @param {Date} d 日期
 * @param {Number} hour 小时
 * @param {Number} min 分钟
 * @param {Number} sec 秒
 * @returns {Date} 一个新的date
 */
export function setDateTime(d: Date, hour: number, min: number, sec: number): Date {
  // eslint-disable-next-line
  const { year, month, date } = getDateObj(d);
  return new Date(year, month, date, hour, min, sec, 0);
}

/**
 * 增加月份
 * @param {Date} date 起始日期
 * @param {Number} num 月份数
 * @returns {Date}
 */
export function subtractMonth(date: Date, num: any): Date {
  const day = date.getDate();
  const newDate = new Date(date);

  let monthNum = num;
  // eslint-disable-next-line no-plusplus
  while (monthNum--) {
    newDate.setDate(0);
  }
  newDate.setDate(day);
  return newDate;
}

/**
 * 减月份
 * @param {Date} date 起始日期
 * @param {Number} num 月份数
 * @returns {Date}
 */
export function addMonth(date: Date, num: number): Date {
  let monthNum = num;
  if (num < 0) monthNum = 0;
  const newDate = new Date(date);
  newDate.setMonth(date.getMonth() + monthNum);
  return newDate;
}

export interface OptionsType {
  firstDayOfWeek: number;
  disableDate: Function;
  minDate: Date;
  maxDate: Date;
  monthText?: string;
}

export function getWeeks(
  { year, month }: { year: number; month: number },
  { firstDayOfWeek, disableDate = () => false, minDate, maxDate }: OptionsType,
) {
  const prependDay = getFirstDayOfMonth({ year, month });

  const appendDay = getLastDayOfMonth({ year, month });

  const maxDays = getDaysInMonth({ year, month });

  const daysArr = [];

  const today = getToday();

  for (let i = 1; i <= maxDays; i++) {
    const currentDay = new Date(year, month, i);

    daysArr.push({
      text: i,
      active: false,
      value: currentDay,
      disabled: disableDate(currentDay) || outOfRanges(currentDay, minDate, maxDate),
      now: isSame(today, currentDay),
      firstDayOfMonth: i === 1,
      lastDayOfMonth: i === maxDays,
      type: 'current-month',
    });
  }

  if (prependDay.getDay() !== firstDayOfWeek - 1) {
    prependDay.setDate(0); // 上一月

    const FLAG = true;
    while (FLAG) {
      daysArr.unshift({
        text: prependDay.getDate().toString(),
        active: false,
        value: new Date(prependDay),
        disabled: disableDate(prependDay) || outOfRanges(prependDay, minDate, maxDate),
        additional: true, // 非当前月
        type: 'prev-month',
      });
      prependDay.setDate(prependDay.getDate() - 1);
      if (prependDay.getDay() === Math.abs(firstDayOfWeek + 6) % 7) break;
    }
  }

  const LEN = 42; // 显示6周

  while (daysArr.length < LEN) {
    appendDay.setDate(appendDay.getDate() + 1);
    daysArr.push({
      text: appendDay.getDate(),
      active: false,
      value: new Date(appendDay),
      disabled: disableDate(appendDay) || outOfRanges(appendDay, minDate, maxDate),
      additional: true, // 非当前月
      type: 'next-month',
    });
  }

  return chunk(daysArr, 7);
}

export function getYears(year: number, { disableDate = () => false, minDate, maxDate }: OptionsType) {
  const startYear = parseInt((year / 10).toString(), 10) * 10;
  const endYear = startYear + 9;

  const yearArr = [];

  const today = getToday();

  for (let i = startYear; i <= endYear; i++) {
    const date = new Date(i, 1);

    let disabledMonth = 0;
    let outOfRangeMonth = 0;

    for (let j = 0; j < 12; j++) {
      const d = new Date(i, j);
      if (disableDate(d)) disabledMonth += 1;
      if (outOfRanges(d, minDate, maxDate)) outOfRangeMonth += 1;
    }

    yearArr.push({
      value: date,
      now: isSame(date, today, 'year'),
      disabled: disabledMonth === 12 || outOfRangeMonth === 12,
      active: false,
      text: `${date.getFullYear()}`,
    });
  }

  return chunk(yearArr, 4);
}

export function getMonths(year: number, { disableDate = () => false, minDate, maxDate, monthText = '' }: OptionsType) {
  const monthArr = [];

  const today = getToday();

  for (let i = 0; i <= 11; i++) {
    const date = new Date(year, i);

    let disabledDay = 0;
    let outOfRangeDay = 0;

    const daysInMonth = getDaysInMonth({ year, month: i });

    for (let j = 1; j <= daysInMonth; j++) {
      const d = new Date(year, i, j);
      if (disableDate(d)) disabledDay += 1;
      if (outOfRanges(d, minDate, maxDate)) outOfRangeDay += 1;
    }

    monthArr.push({
      value: date,
      now: isSame(date, today, 'month'),
      disabled: disabledDay === daysInMonth || outOfRangeDay === daysInMonth,
      active: false,
      text: `${date.getMonth() + 1} ${monthText}`,
    });
  }

  return chunk(monthArr, 4);
}

interface DateTime {
  active: boolean;
  highlight: boolean;
  startOfRange: boolean;
  endOfRange: boolean;
  value: Date;
}

export function flagActive(data: any[], { ...args }: any) {
  const { start, end, type = 'date' } = args;

  if (!end) {
    return data.map((row: any[]) =>
      row.map((item: DateTime) => {
        Object.assign(item, { active: isSame(item.value, start, type) });
        return item;
      }),
    );
  }

  return data.map((row: any[]) =>
    row.map((item: DateTime) => {
      const date = item.value;
      const isStart = isSame(start, date, type);
      const isEnd = isSame(end, date, type);
      Object.assign(item, {
        active: isStart || isEnd,
        highlight: isBetween(date, { start, end }),
        startOfRange: isStart,
        endOfRange: isEnd,
      });
      return item;
    }),
  );
}
