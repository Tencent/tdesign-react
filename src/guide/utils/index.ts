import { scrollToParentVisibleArea } from './getScrollParent';
import getRelativePosition from './getRelativePosition';
import getTargetElm from './getTargetElm';
import scrollToElm from './scrollToElm';
import useMount from './useMount';

export { scrollToParentVisibleArea, getRelativePosition, getTargetElm, scrollToElm, useMount };
