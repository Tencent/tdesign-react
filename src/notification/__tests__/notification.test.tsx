import React from 'react';
import { testExamples, render, fireEvent } from '@test/utils';
import { Notification, NotificationPlugin } from 'tdesign-react';

testExamples(__dirname);

describe('Notification test', () => {
  test('mount and unmount', () => {
    const wrapper = render(<Notification />);

    expect(() => wrapper.unmount()).not.toThrow();
  });

  test('set props', async () => {
    const titleText = '标题';
    const contentId = 'content';
    const contentText = '内容';
    const closeBtnId = 'close';
    const closeBtnText = '关闭';
    const footerId = 'footer';

    const wrapper = render(
      <Notification
        title={titleText}
        content={<div data-testid={contentId}>{contentText}</div>}
        theme="success"
        icon={<div>test</div>}
        closeBtn={<div data-testid={closeBtnId}>{closeBtnText}</div>}
        footer={
          <div data-testid={footerId} className="t-notification__detail">
            <span className="t-notification__detail-item">确定</span>
          </div>
        }
      />,
    );

    /**
     * @desc 更改 title 的断言测试。新的 api title 只接受纯字符串，需要修改测试方法
     */
    const title = document.querySelectorAll('.t-notification__title');
    expect(title.length).toBe(1);
    expect(title[0].innerHTML).toBe(titleText);

    expect(wrapper.queryByTestId(contentId)).toHaveTextContent(contentText);
    expect(wrapper.queryByTestId(closeBtnId)).toHaveTextContent(closeBtnText);
    expect(wrapper.queryByTestId(footerId)).not.toBeNull();
  });

  test('open and close', async () => {
    const notification = await NotificationPlugin.info({});
    expect(document.querySelectorAll('.t-notification').length).toBe(1);
    notification.close();
    expect(document.querySelectorAll('.t-notification').length).toBe(0);

    const notificationPromise = NotificationPlugin.info({});

    await notificationPromise;

    expect(document.querySelectorAll('.t-notification').length).toBe(1);
    await NotificationPlugin.close(notificationPromise);
    expect(document.querySelectorAll('.t-notification').length).toBe(0);
  });

  test('open with theme', async () => {
    NotificationPlugin.closeAll();

    await NotificationPlugin.info({});
    await NotificationPlugin.success({});
    await NotificationPlugin.warning({});
    await NotificationPlugin.error({});

    expect(document.querySelectorAll('.t-notification .t-is-info').length).toBe(1);
    expect(document.querySelectorAll('.t-notification .t-is-success').length).toBe(1);
    expect(document.querySelectorAll('.t-notification .t-is-warning').length).toBe(1);
    expect(document.querySelectorAll('.t-notification .t-is-error').length).toBe(1);
  });

  test('open with placement', async () => {
    NotificationPlugin.closeAll();

    await NotificationPlugin.info({ placement: 'top-left' });
    await NotificationPlugin.info({ placement: 'top-right' });
    await NotificationPlugin.info({ placement: 'bottom-left' });
    await NotificationPlugin.info({ placement: 'bottom-right' });

    expect(document.querySelectorAll('.t-notification__show--top-left').length).toBe(1);
    expect(document.querySelectorAll('.t-notification__show--top-right').length).toBe(1);
    expect(document.querySelectorAll('.t-notification__show--bottom-left').length).toBe(1);
    expect(document.querySelectorAll('.t-notification__show--bottom-right').length).toBe(1);
  });

  test('auto close', async () => {
    NotificationPlugin.closeAll();

    await NotificationPlugin.info({ duration: 3000 });

    expect(document.querySelectorAll('.t-notification').length).toBe(1);

    jest.runAllTimers();

    expect(document.querySelectorAll('.t-notification').length).toBe(0);
  });

  test('click close button', async () => {
    NotificationPlugin.closeAll();

    await NotificationPlugin.info({
      closeBtn: <span id="close_button">关闭</span>,
    });

    fireEvent.click(document.querySelector('#close_button'));

    expect(document.querySelectorAll('.t-notification').length).toBe(0);
  });
});
