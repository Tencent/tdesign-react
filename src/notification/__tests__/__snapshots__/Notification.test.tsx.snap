// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-notification t-is-info"
  >
    <div
      class="t-notification__icon"
    >
      <svg
        class="t-icon t-icon-info-circle-filled t-is-info"
        fill="none"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
      >
        <path
          d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
          fill="currentColor"
          fill-opacity="0.9"
        />
      </svg>
    </div>
    <div
      class="t-notification__main"
    >
      <div
        class="t-notification__title__wrap"
      >
        <span
          class="t-notification__title"
        >
          标题名称
        </span>
      </div>
      <div
        class="t-notification__content"
      >
        这是一条消息通知
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`close-all.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-row"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        点击打开多个消息
      </span>
    </button>
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        点击关闭多个消息
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`content.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-columns"
  >
    <div
      class="t-notification t-is-info"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-info"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            自定义内容（字符串）
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          这是一条消息通知
        </div>
      </div>
    </div>
    <div
      class="t-notification t-is-info"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-info"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            自定义内容
          </span>
        </div>
        <div>
          这是一条消息通知
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`footer.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-notification t-is-info"
  >
    <div
      class="t-notification__icon"
    >
      <svg
        class="t-icon t-icon-info-circle-filled t-is-info"
        fill="none"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
      >
        <path
          d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
          fill="currentColor"
          fill-opacity="0.9"
        />
      </svg>
    </div>
    <div
      class="t-notification__main"
    >
      <div
        class="t-notification__title__wrap"
      >
        <span
          class="t-notification__title"
        >
          自定义底部详情
        </span>
      </div>
      <div
        class="t-notification__content"
      >
        这是一条消息通知
      </div>
      <div
        class="t-notification__detail"
      >
        <div
          class="t-notification__detail"
        >
          <span
            class="t-notification__detail-item"
          >
            重启
          </span>
          <span
            class="t-notification__detail-item t-is-active"
          >
            查看详情
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`icon.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-column"
  >
    <div
      class="t-notification t-is-info"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-info"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            普通通知
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知
        </div>
      </div>
    </div>
    <div
      class="t-notification t-is-error"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-error"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            危险通知
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知
        </div>
      </div>
    </div>
    <div
      class="t-notification t-is-warning"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-warning"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            告警通知
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知
        </div>
      </div>
    </div>
    <div
      class="t-notification t-is-success"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-check-circle-filled t-is-success"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM4.5 8.2l.7-.7L7 9.3l3.8-3.8.7.7L7 10.7 4.5 8.2z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            成功通知
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`operation.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-column"
  >
    <div
      class="t-notification t-is-info"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-info"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            超出的文本省略号显示
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          文案不限长度，但是展示最大显示三行折行的末尾显示折行末尾显示折行末尾显示折行末尾显示折行末尾显示折行折行末尾显示折行折行末尾显示折行末尾显示折行折行末尾
        </div>
        <div
          class="t-notification__detail"
        >
          <div>
            <button
              class="t-button t-button--theme-primary t-button--variant-text"
              type="button"
            >
              <span
                class="t-button__text"
              >
                查看详情
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="t-notification t-is-info"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-info"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            自定义底部
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          使用 props function 自定义底部内容
        </div>
        <div
          class="t-notification__detail"
        >
          <div>
            <button
              class="t-button t-button--theme-primary t-button--variant-text"
              type="button"
            >
              <span
                class="t-button__text"
              >
                查看详情
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="t-notification t-is-info"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-info"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            <div>
              自定义标题 
              <small>
                我是副标题
              </small>
            </div>
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          1. 使用 props function 自定义标题；2. 使用插槽自定义底部内容
        </div>
        <div
          class="t-notification__detail"
        >
          <div
            slot="footer"
          >
            <button
              class="t-button t-button--theme-default t-button--variant-text"
              type="button"
            >
              <span
                class="t-button__text"
              >
                知道了
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="t-notification t-is-info"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-info"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            <div>
              自定义标题 
              <small>
                我是副标题
              </small>
            </div>
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          1. 使用插槽自定义标题 2. 使用插槽自定义底部内容
        </div>
        <div
          class="t-notification__detail"
        >
          <div>
            <button
              class="t-button t-button--theme-default t-button--variant-text"
              type="button"
            >
              <span
                class="t-button__text"
              >
                重启
              </span>
            </button>
            <button
              class="t-button t-button--theme-primary t-button--variant-text"
              type="button"
            >
              <span
                class="t-button__text"
              >
                稍后提醒我(10s)
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="t-notification t-is-info"
    >
      <div
        class="t-notification__icon"
      >
        <svg
          class="t-icon t-icon-info-circle-filled t-is-info"
          fill="none"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
        >
          <path
            d="M8 15A7 7 0 108 1a7 7 0 000 14zM7.4 4h1.2v1.2H7.4V4zm.1 2.5h1V12h-1V6.5z"
            fill="currentColor"
            fill-opacity="0.9"
          />
        </svg>
      </div>
      <div
        class="t-notification__main"
      >
        <div
          class="t-notification__title__wrap"
        >
          <span
            class="t-notification__title"
          >
            自定义内容
          </span>
        </div>
        <div
          class="t-notification__content"
        >
          使用插槽自定义内容
        </div>
        <div
          class="t-notification__detail"
        >
          <div>
            <button
              class="t-button t-button--theme-default t-button--variant-text"
              type="button"
            >
              <span
                class="t-button__text"
              >
                重启
              </span>
            </button>
            <button
              class="t-button t-button--theme-primary t-button--variant-text"
              type="button"
            >
              <span
                class="t-button__text"
              >
                更多
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`placement.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-column"
  >
    <div
      class="tdesign-demo-block-row"
    >
      <div
        class="t-input__wrap"
        style="width: 130px; display: inline-block; margin: 0px 20px 36px 0px;"
        value=""
      >
        <div
          class="t-input t-size-m t-align-left t-is-default"
        >
          <input
            class="t-input__inner"
            placeholder="请输入横向偏移量"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="t-input__wrap"
        style="width: 130px; display: inline-block; margin: 0px 20px 36px 0px;"
        value=""
      >
        <div
          class="t-input t-size-m t-align-left t-is-default"
        >
          <input
            class="t-input__inner"
            placeholder="请输入纵向偏移量"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="tdesign-demo-block-row"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          左上角
        </span>
      </button>
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          右上角
        </span>
      </button>
    </div>
    <div
      class="tdesign-demo-block-row"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          左下角
        </span>
      </button>
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          右下角
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`plugin.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-row"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        信息
      </span>
    </button>
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        成功
      </span>
    </button>
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        警告
      </span>
    </button>
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        错误
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`toggle.jsx 1`] = `
<DocumentFragment>
  <button
    class="t-button t-button--theme-primary t-button--variant-base"
    type="button"
  >
    <span
      class="t-button__text"
    >
      自由控制关闭时机
    </span>
  </button>
</DocumentFragment>
`;
